util/src/main/java/io/kubernetes/client/util/authenticators/OpenIDConnectAuthenticator.java

```diff
@@ -0,0 +1,246 @@
+          }
+        } else {
+          ks.setCertificateEntry(alias, c.iterator().next());
+        }
+
+        TrustManagerFactory tmf = TrustManagerFactory.getInstance("PKIX");
+        tmf.init(ks);
+
+        sslContext = SSLContext.getInstance("TLSv1.2");
+        sslContext.init(null, tmf.getTrustManagers(), new SecureRandom());
+
+      } catch (KeyStoreException
+          | NoSuchAlgorithmException
+          | CertificateException
+          | IOException
+          | KeyManagementException e) {
+        throw new RuntimeException("Could not import idp certificate", e);
+      }
+    }
+
+    if (clientSecret == null) {
+      clientSecret = "";
+    }
+
+    StringBuilder sb = new StringBuilder();
+    sb.append(issuer);
+    if (!issuer.endsWith("/")) {
+      sb.append("/");
+    }
+    sb.append(".well-known/openid-configuration");
+
+    String wellKnownURL = sb.toString();
+
+    try {
+      URL wellKnown = new URL(wellKnownURL);
+      HttpsURLConnection https = (HttpsURLConnection) wellKnown.openConnection();
+      https.setRequestMethod("GET");
+      if (sslContext != null) {
+        https.setSSLSocketFactory(sslContext.getSocketFactory());
+      }
+      https.setUseCaches(false);
+      int code = https.getResponseCode();
+      if (code == 200) {
+        Scanner scanner = new Scanner(https.getInputStream(), StandardCharsets.UTF_8.name());
+        String json = scanner.useDelimiter("\\A").next();
+
+        JSONObject wellKnownJson = (JSONObject) new JSONParser().parse(json);
+        String tokenUrl = (String) wellKnownJson.get("token_endpoint");
+
+        URL tokenEndpoint = new URL(tokenUrl);
+        https = (HttpsURLConnection) tokenEndpoint.openConnection();
+        https.setRequestMethod("POST");
+        if (sslContext != null) {
+          https.setSSLSocketFactory(sslContext.getSocketFactory());
+        }
+
+        String credentials =
+            Base64.getEncoder()
+                .encodeToString(
+                    new StringBuilder()
+                        .append(clientId)
+                        .append(':')
+                        .append(clientSecret)
+                        .toString()
+                        .getBytes("UTF-8"));
+        https.setRequestProperty(
+            "Authorization", new StringBuilder().append("Bearer ").append(credentials).toString());
+        https.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");
+        https.setDoOutput(true);
+
+        String urlData =
+            new StringBuilder()
+                .append("refresh_token=")
+                .append(URLEncoder.encode(refreshToken, "UTF-8"))
+                .toString();
+        OutputStream ou = https.getOutputStream();
+        ou.write(urlData.getBytes("UTF-8"));
+        ou.flush();
+        ou.close();
+
+        code = https.getResponseCode();
+
+        if (code == 200) {
+          scanner = new Scanner(https.getInputStream(), StandardCharsets.UTF_8.name());
```

Comment:
a bunch of this code is duplicated from above, and this method is pretty long anyway, refactor the HTTP object loading logic out into a helper method.